shared-bindings/_canio/Message.c

```diff
@@ -0,0 +1,215 @@
+        mp_raise_TypeError(translate("specify size or data, but not both"));
+    }
+
+    mp_buffer_info_t data;
+    if (specified_data) {
+        mp_get_buffer_raise(args[ARG_data].u_obj, &data, MP_BUFFER_READ);
+    } else if (specified_size) {
+        data.buf = 0;
+        data.len = size;
+    } else {
+        data.buf = 0;
+        data.len = 0;
+    }
+
+    if(data.len > 8) {
+        mp_raise_ValueError(translate("Messages limited to 8 bytes"));
+    }
+
+    canio_message_obj_t *self = m_new_obj(canio_message_obj_t);
+    self->base.type = &canio_message_type;
+    common_hal_canio_message_construct(self, args[ARG_id].u_int, data.buf, data.len, rtr);
+    return self;
+}
+
+//|     id: int
+//|     """The numeric ID of the message"""
+//|
+STATIC mp_obj_t canio_message_id_get(const mp_obj_t self_in) {
+    canio_message_obj_t *self = self_in;
+    return MP_OBJ_NEW_SMALL_INT(common_hal_canio_message_id_get(self));
+}
+MP_DEFINE_CONST_FUN_OBJ_1(canio_message_id_get_obj, canio_message_id_get);
+
+STATIC mp_obj_t canio_message_id_set(const mp_obj_t self_in, const mp_obj_t id) {
+    canio_message_obj_t *self = self_in;
+    common_hal_canio_message_id_set(self, mp_obj_get_int(id));
+    return mp_const_none;
+}
+MP_DEFINE_CONST_FUN_OBJ_2(canio_message_id_set_obj, canio_message_id_set);
+
+STATIC const mp_obj_property_t canio_message_id_obj = {
+    .base.type = &mp_type_property,
+    .proxy = {(mp_obj_t)&canio_message_id_get_obj,
+              (mp_obj_t)&canio_message_id_set_obj,
+              (mp_obj_t)&mp_const_none_obj},
+};
+
+//|     data: bytes
+//|     """The content of the message, or dummy content in the case of an rtr.
+//|
+//|     Assigning to data also sets the length."""
+//|
+STATIC mp_obj_t canio_message_data_get(const mp_obj_t self_in) {
+    canio_message_obj_t *self = self_in;
+    return mp_obj_new_bytes((const byte*)common_hal_canio_message_data_get(self), common_hal_canio_message_size_get(self));
+}
+MP_DEFINE_CONST_FUN_OBJ_1(canio_message_data_get_obj, canio_message_data_get);
+
+STATIC mp_obj_t canio_message_data_set(const mp_obj_t self_in, const mp_obj_t data_in) {
+    canio_message_obj_t *self = self_in;
+    mp_buffer_info_t data;
+    mp_get_buffer_raise(data_in, &data, MP_BUFFER_READ);
+    if(data.len > 8) {
+        mp_raise_ValueError(translate("Messages limited to 8 bytes"));
+    }
+    common_hal_canio_message_data_set(self, data.buf, data.len);
+    return mp_const_none;
+}
+MP_DEFINE_CONST_FUN_OBJ_2(canio_message_data_set_obj, canio_message_data_set);
+
+
+STATIC const mp_obj_property_t canio_message_data_obj = {
+    .base.type = &mp_type_property,
+    .proxy = {(mp_obj_t)&canio_message_data_get_obj,
+              (mp_obj_t)&canio_message_data_set_obj,
+              (mp_obj_t)&mp_const_none_obj},
+};
+
+
+//|     size: int
+//|     """The length of the message, or the length of the requested data in the case of an rtr
+//|
+//|     Assigning to the length sets all the data bytes to zero"""
```

Comment:
"Remote Transmission Request" (RTR) messages have a size but no meaningful data: `rtr_message = Message(id=0x408, rtr=True, size=8)`

Non-RTR messages have meaningful (hopefully) data: `std_message = Message(id=0x408, data=b'adaf00')`

For readinto (which can result in an RTR message or a non-RTR message), one Message class has to allow both options.  So that constructing an RTR message didn't require allocating a dummy bytes() object of the desired length, I made `data` and `size` (could be `length`?) both available as properties and constructor parameters.

(A message object doesn't work with `len()` and can't be subscripted or sliced, you have to retrieve the .data to do that)

Reply:
It seems to me that having a separate RemoteTransmissionRequest object could make things simpler. We'd need to drop `readinto` then though.

Alternatively, what if we reused an RTR message to reply (I assume that it is initiating a reply)? We remove size and have an empty bytearray as data. That way you fill in the bytearray, clear rtr and send it back.

Reply:
Yeah, if readinto gives you a message with the RTR flag set, you set data, clear RTR (hmmm .. so setting data should clear RTR?) and send it back.

Reply:
I think there is a question whether `.data` is mutable or not. You can't automatically clear RTR if you mutate part of `.data`

Reply:
The `Message.data` property returns an immutable bytes copy of the data, so you can't mutate it.